File: Cpp_Parallel/src/lib/ParallelBubbleSort.cpp
#include "include/ParallelBubbleSort.h"

void ParallelBubbleSort::parallelBubbleSort(std::vector<int>& array, int numThreads) {
    int n = array.size();
    std::vector<std::thread> threads;

    for (int i = 0; i < n - 1; i++) {
        int outerIndex = i;
        threads.emplace_back([=, &array]() {
            bubbleSortSegment(array, 0, n - outerIndex - 1);
        });
    }

    for (auto& thread : threads) {
        thread.join();
    }
}

void ParallelBubbleSort::bubbleSortSegment(std::vector<int>& array, int start, int end) {
    for (int j = start; j < end; j++) {
        if (array[j] > array[j + 1]) {
            std::swap(array[j], array[j + 1]);
        }
    }
}

File: Java_Parallel/src/ParallelSorts/ParallelBubbleSort.java
package ParallelSorts;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ParallelBubbleSort {

    public static void parallelBubbleSort(int[] array, int numThreads) {
        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
        int n = array.length;

        for (int i = 0; i < n - 1; i++) {
            final int outerIndex = i;
            executor.submit(() -> {
                for (int j = 0; j < n - outerIndex - 1; j++) {
                    if (array[j] > array[j + 1]) {
                        swap(array, j, j + 1);
                    }
                }
            });
        }
        executor.shutdown();
        while (!executor.isTerminated()) {} // Wait for all tasks to finish
    }

    private static void swap(int[] array, int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
}
